### PR TITLE
StreamingCredentialProvider support

### PR DESCRIPTION
**Pull request by vladvildanov**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [ ] Do tests and lints pass with this change?
- [ ] Do the CI tests pass with this change (enable it first in your forked repo and wait for the github action build to finish)?
- [ ] Is the new or changed code fully tested?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [ ] Is there an example added to the examples folder (if applicable)?
- [ ] Was the change added to CHANGES file?

_NOTE: these things are not required to open a PR and can be done
afterwards / while the PR is open._

### Description of change

_Please provide a description of the change here._

### CODE REVIEW

**Review by uglide — CHANGES_REQUESTED** [review]

File: redis/asyncio/client.py [hunk 1]
```diff
@@ -233,6 +235,7 @@ def __init__(
         redis_connect_func=None,
         credential_provider: Optional[CredentialProvider] = None,
         protocol: Optional[int] = 2,
+        event_dispatcher: Optional[EventDispatcher] = EventDispatcher(),
```
[the remark this answers is not in the document]
Comment (vladvildanov) [reply redis/asyncio/client.py:238]: But why it's a problem? Dispatcher just holds a connection between event and associated listener, what is the point to have multiple dispatchers?

File: redis/auth/token_manager.py [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+    def get_retry_policy(self) -> RetryPolicy:
+        """
+        Represents the retry policy for token requests.
+
+        :return: RetryPolicy
+        """
+        return self._retry_policy
+
+
+class TokenManager:
```
[the remark this answers is not in the document]
Comment (vladvildanov) [reply redis/auth/token_manager.py:114]: Good idea to add a logger, will do it. But actually the exceptions from IDP, are handled. IDP is one method only and it's wrapped by try catch in both sync and async calls

https://github.com/redis/redis-py/blob/vv-tba-support/redis/auth/token_manager.py#L186
https://github.com/redis/redis-py/blob/vv-tba-support/redis/auth/token_manager.py#L200

File: redis/auth/token_manager.py [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+                    raise TokenRenewalErr(e)
+
+            if delay <= 0:
+                return
+
+            loop = asyncio.get_running_loop()
+            self._next_timer = loop.call_later(delay, self._renew_token)
+            return token_res
+        except Exception as e:
+            if self._listener.on_error is None:
```
[the remark this answers is not in the document]
Comment (vladvildanov) [reply redis/auth/token_manager.py:262]: Yeah, there's no place for now where I set this callback, but this check is doing exactly what you mention it checks if callback is set and if not it's just raises an exception that will be propagated to upper layers

File: redis/asyncio/client.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (uglide) [reply redis/asyncio/client.py:238]: The problem with single shared Dispatcher is that you also share all event listeners. As a consequence all event listeners should be thread-safe, which is not the case for many of them currently.

In the code like this:
```
    def listen(self, event: AfterSingleConnectionInstantiationEvent):
        if isinstance(event.connection.credential_provider, StreamingCredentialProvider):
            self._event = event

            if event.client_type == ClientType.SYNC:
                event.connection.credential_provider.on_next(self._re_auth)
            else:
                event.connection.credential_provider.on_next(self._re_auth_async)

    def _re_auth(self, token):
        with self._event.connection_lock:
            self._event.connection.send_command('AUTH', token.try_get('oid'), token.get_value())
            self._event.connection.read_response()
```
Multiple threads/clients will override the `self._event`, and random clients will receive tokens from random credential providers, **which is a security vulnerability.**

So let's follow the typical approach with:

```
event_dispatcher: Optional[EventDispatcher] = None,
# ...
if event_dispatcher is None:
    self._event_dispatcher = EventDispatcher()
```

File: redis/auth/token_manager.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (uglide) [reply redis/auth/token_manager.py:269]: Do you have a test for this to ensure that this is precisely what happens with the sync client? If not, I suggest adding it.

File: redis/asyncio/client.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (vladvildanov) [reply redis/asyncio/client.py]: Got fixed now

**Review by uglide — APPROVED** [review]
LGTM

**Pull request merged**